File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/gatt/client/BleHrClientTest.kt
package com.polar.androidcommunications.api.ble.model.gatt.client

import com.polar.androidcommunications.api.ble.model.gatt.BleGattTxInterface
import com.polar.androidcommunications.api.ble.model.gatt.client.BleHrClient.Companion.HR_MEASUREMENT
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runTest
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.util.*

class BleHrClientTest {

    @MockK
    private lateinit var txInterface: BleGattTxInterface

    private lateinit var bleHrClient: BleHrClient

    @Before
    fun setUp() {
        MockKAnnotations.init(this)
        bleHrClient = BleHrClient(txInterface)
        every { txInterface.isConnected() } returns true
        every { txInterface.setCharacteristicNotify(any(), any(), any()) } returns Unit
    }

    @Test
    fun `hr format uint8`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true

        // HEX: 00 FF
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x00
        // Heart rate value format bit:     0 (uint8)
        // 1:       Heart Rate Measurement Value field      size 1:     0xFF
        val expectedHeartRate1 = 255
        val measurementFrame1 = byteArrayOf(0x00.toByte(), 0xFF.toByte())
        // HEX: 00 7F
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x00
        // Heart rate value format bit:     0 (uint8)
        // 1:       Heart Rate Measurement Value field      size 1:     0x7F
        val expectedHeartRate2 = 127
        val measurementFrame2 = byteArrayOf(0x00.toByte(), 0x7F.toByte())

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame2, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(2, values.size)
        assertEquals(expectedHeartRate1, values[0].hrValue)
        assertEquals(expectedHeartRate2, values[1].hrValue)
    }

    @Test
    fun `hr format uint16`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true
        // HEX: 01 80 80
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x01
        // Heart rate value format bit      1 (uint16)
        // 1..2:    Heart Rate Measurement Value field      size 2:     0x80 0x80 (32896)
        val expectedHeartRate1 = 32896
        val measurementFrame1 = byteArrayOf(0x01.toByte(), 0x80.toByte(), 0x80.toByte())
        // HEX: 01 7F 7F
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x01
        // Heart rate value format bit      1 (uint16)
        // 1..2:    Heart Rate Measurement Value field      size 2:     0x7F 0x7F
        val expectedHeartRate2 = 32639
        val measurementFrame2 = byteArrayOf(0x01.toByte(), 0x7F.toByte(), 0x7F.toByte())

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame2, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(2, values.size)
        assertEquals(expectedHeartRate1, values[0].hrValue)
        assertEquals(expectedHeartRate2, values[1].hrValue)
    }

    @Test
    fun `sensor contact supported`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true
        // HEX: 06 00
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x06
        // Sensor Contact Status bit        1
        // Sensor Contact Support bit       1 (Supported)
        val measurementFrame0 = byteArrayOf(0x06.toByte(), 0x00.toByte())
        // HEX: 04 00
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x04
        // Sensor Contact Status bit        0
        // Sensor Contact Support bit       1 (Supported)
        val measurementFrame1 = byteArrayOf(0x04.toByte(), 0x00.toByte())

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame0, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(2, values.size)
        assertTrue(values[0].sensorContact)
        assertTrue(values[0].sensorContactSupported)
        assertFalse(values[1].sensorContact)
        assertTrue(values[1].sensorContactSupported)
    }

    @Test
    fun `sensor contact not supported`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true

        // HEX: 02 00
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x02
        // Sensor Contact Status bit        1
        // Sensor Contact Support bit       0 (Not supported)
        val measurementFrame1 = byteArrayOf(0x02.toByte(), 0x00.toByte())

        // HEX: 00 00
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x00
        // Sensor Contact Status bit        0
        // Sensor Contact Support bit       0 (Not supported)
        val measurementFrame2 = byteArrayOf(0x00.toByte(), 0x00.toByte())

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame2, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(2, values.size)
        assertFalse(values[0].sensorContactSupported)
        assertFalse(values[1].sensorContactSupported)
    }

    @Test
    fun `energy expended`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true

        // HEX: 09 00 00 FF FF
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x09
        // Heart rate value format bit      1 (uint16)
        // Energy Expended Status bit       1 (Energy Expended field is present)
        // 1..2:    Heart Rate Measurement Value field      size 2:     0x00 0x00
        // 3..4:    Energy Expended field                   size 2:     0xFF 0xFF
        val expectedHeartRate1 = 0
        val energyExpended1 = 65535
        val measurementFrame1 =
            byteArrayOf(0x09.toByte(), 0x00.toByte(), 0x00.toByte(), 0xFF.toByte(), 0xFF.toByte())
        // HEX: 08 00 7F 80
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x08
        // Heart rate value format bit      0 (uint8)
        // Energy Expended Status bit       1 (Energy Expended field is present)
        // 1:       Heart Rate Measurement Value field      size 1:     0x00
        // 2..3:    Energy Expended field                   size 2:     0x7F 0x80
        val expectedHeartRate2 = 0
        val energyExpended2 = 32895
        val measurementFrame2 =
            byteArrayOf(0x08.toByte(), 0x00.toByte(), 0x7F.toByte(), 0x80.toByte())

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame2, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(2, values.size)
        assertEquals(expectedHeartRate1, values[0].hrValue)
        assertEquals(energyExpended1, values[0].energy)
        assertEquals(expectedHeartRate2, values[1].hrValue)
        assertEquals(energyExpended2, values[1].energy)
    }

    @Test
    fun `rr interval`() = runTest {
        //Arrange
        val characteristic: UUID = HR_MEASUREMENT
        val status = 0
        val notifying = true
        // HEX: 10 00 FF FF
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x10
        // Heart rate value format bit      0 (uint8)
        // RR-interval bit                  1 (RR-Interval values are present)
        // 1:       Heart Rate Measurement Value field      size 1:     0x00
        // 2..3:    RR-Interval Field                       size 2:     0xFF 0xFF
        val expectedHeartRate1 = 0
        val sample1ExpectedRRin1024Unit = 65535
        val sample1ExpectedRRinMsUnit = 63999
        val measurementFrame1 =
            byteArrayOf(0x10.toByte(), 0x00.toByte(), 0xFF.toByte(), 0xFF.toByte())
        // HEX: 11 00 00 FF FF 7F 80
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x11
        // Heart rate value format bit      1 (uint16)
        // RR-interval bit                  1 (RR-Interval values are present)
        // 1..2:    Heart Rate Measurement Value field      size 1:     0x00 0x00
        // 3..4:    RR-Interval Field                       size 2:     0xFF 0xFF
        // 5..6:    RR-Interval Field                       size 2:     0x7F 0x80
        val expectedHeartRate2 = 0
        val sample2ExpectedRRin1024Unit = 65535
        val sample2ExpectedRRinMsUnit = 63999
        val sample3ExpectedRRin1024Unit = 32895
        val sample3ExpectedRRinMsUnit = 32124
        val measurementFrame2 = byteArrayOf(
            0x11.toByte(),
            0x00.toByte(),
            0x00.toByte(),
            0xFF.toByte(),
            0xFF.toByte(),
            0x7F.toByte(),
            0x80.toByte()
        )

        // HEX: 10 00 00 00
        // index    type                                                data:
        // 0:       Flags field                             size 1:     0x10
        // Heart rate value format bit      0 (uint8)
        // RR-interval bit                  1 (RR-Interval values are present)
        // 1:       Heart Rate Measurement Value field      size 1:     0x00
        // 2..3:    RR-Interval Field                       size 2:     0x00 0x00
        val expectedHeartRate3 = 0
        val sample4ExpectedRRin1024Unit = 0
        val sample4ExpectedRRinMsUnit = 0
        val measurementFrame3 = byteArrayOf(
            0x11.toByte(),
            0x00.toByte(),
            0x00.toByte(),
            0x00.toByte(),
            0x00.toByte()
        )

        //Act
        val values = mutableListOf<BleHrClient.HrNotificationData>()
        val job = launch { bleHrClient.observeHrNotifications(true).collect { values.add(it) } }
        testScheduler.advanceUntilIdle()
        bleHrClient.processServiceData(characteristic, measurementFrame1, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame2, status, notifying)
        bleHrClient.processServiceData(characteristic, measurementFrame3, status, notifying)
        testScheduler.advanceUntilIdle()
        job.cancel()

        //Assert
        assertEquals(3, values.size)

        val hrData1 = values[0]
        assertEquals(expectedHeartRate1, hrData1.hrValue)
        assertEquals(1, hrData1.rrs.size)
        assertEquals(1, hrData1.rrsMs.size)
        assertTrue(hrData1.rrPresent)
        assertEquals(sample1ExpectedRRin1024Unit, hrData1.rrs[0])
        assertEquals(sample1ExpectedRRinMsUnit, hrData1.rrsMs[0])

        val hrData2 = values[1]
        assertEquals(expectedHeartRate2, hrData2.hrValue)
        assertEquals(2, hrData2.rrs.size)
        assertEquals(2, hrData2.rrsMs.size)
        assertTrue(hrData2.rrPresent)
        assertEquals(sample2ExpectedRRin1024Unit, hrData2.rrs[0])
        assertEquals(sample2ExpectedRRinMsUnit, hrData2.rrsMs[0])

        assertEquals(sample3ExpectedRRin1024Unit, hrData2.rrs[1])
        assertEquals(sample3ExpectedRRinMsUnit, hrData2.rrsMs[1])

        val hrData3 = values[2]
        assertEquals(expectedHeartRate3, hrData3.hrValue)
        assertEquals(1, hrData3.rrs.size)
        assertEquals(1, hrData3.rrsMs.size)
        assertTrue(hrData3.rrPresent)
        assertEquals(sample4ExpectedRRin1024Unit, hrData3.rrs[0])
        assertEquals(sample4ExpectedRRinMsUnit, hrData3.rrsMs[0])
    }
}

File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/BleHrClientTest.swift
//  Copyright © 2021 Polar. All rights reserved.

import XCTest
@testable import iOSCommunications
import CoreBluetooth

class BleHrClientTest: XCTestCase {

    var bleHrClient: BleHrClient!
    var mockGattServiceTransmitterImpl: MockPolarGattServiceTransmitter!

    override func setUpWithError() throws {
        mockGattServiceTransmitterImpl = MockPolarGattServiceTransmitter()
        bleHrClient = BleHrClient(gattServiceTransmitter: mockGattServiceTransmitterImpl)
    }

    override func tearDownWithError() throws {
        mockGattServiceTransmitterImpl = nil
        bleHrClient = nil
    }

    // MARK: - Helpers

    /// Collects up to `count` items from a stream, then cancels it.
    private func collect(_ count: Int, from stream: AsyncThrowingStream<BleHrClient.BleHrNotification, Error>) async throws -> [BleHrClient.BleHrNotification] {
        var results: [BleHrClient.BleHrNotification] = []
        for try await item in stream {
            results.append(item)
            if results.count == count { break }
        }
        return results
    }

    // GIVEN that BLE HR Service client observable is subscribed
    // WHEN the connection to device is missing
    // THEN stream should immediately throw gattDisconnected
    func testConnectionToDeviceIsLost() async throws {
        // Arrange
        mockGattServiceTransmitterImpl.mockConnectionStatus = false

        // Act & Assert
        let stream = bleHrClient.observeHrNotifications(true)
        do {
            for try await _ in stream { XCTFail("Should not receive data") }
            XCTFail("Should have thrown an error")
        } catch {
            guard case BleGattException.gattDisconnected = error else {
                return XCTFail("Expected gattDisconnected, got \(error)")
            }
        }
    }

    // GIVEN that BLE HR Service client receives heart rate data updates
    // WHEN hr data stream is being consumed
    // THEN at some point device connection is lost
    func testDeviceDisconnectsWhileStreaming() async throws {
        // Arrange
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)

        // Use a Task that returns its collected results and error so Swift value-
        // semantics don't hide mutations made inside the @Sendable closure.
        let task = Task<(items: [BleHrClient.BleHrNotification], error: Error?), Never> {
            var collected: [BleHrClient.BleHrNotification] = []
            do {
                for try await item in stream { collected.append(item) }
            } catch {
                return (collected, error)
            }
            return (collected, nil)
        }

        try await Task.sleep(nanoseconds: 20_000_000)

        // Act
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 1]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 2]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 3]), err: 0)
        bleHrClient.disconnected()

        let (results, thrownError) = await task.value

        // Assert — 3 HR values then a disconnection error
        XCTAssertEqual(3, results.count)
        for item in results { XCTAssertTrue((1...3).contains(item.hr)) }
        XCTAssertNotNil(thrownError)
        guard case BleGattException.gattDisconnected = thrownError! else {
            return XCTFail("Expected gattDisconnected, got \(thrownError!)")
        }
    }

    // GIVEN that BLE HR Service client observable is receiving hr values
    // WHEN stream is subscribed
    // THEN no values in cache
    func testNoCachedValues() async throws {
        // Arrange
        let characteristic = CBUUID(string: "2a37")

        // Emit first value BEFORE subscribing — should NOT be cached
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 101]), err: 0)

        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(1, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        // Emit second value AFTER subscribing — should be received
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 102]), err: 0)

        let results = try await task.value

        // Assert — only the second value should be received
        XCTAssertEqual(1, results.count)
        XCTAssertEqual(102, results[0].hr)
    }

    // GIVEN BLE HR Service client receives heart rate value in uint8 data format
    // WHEN heart rate stream is consumed
    // THEN the values are emitted
    func testHrFormatUint8() async throws {
        // Arrange
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(2, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        // Act
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 0xFF]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 0x7F]), err: 0)

        let results = try await task.value

        // Assert
        XCTAssertEqual(2, results.count)
        XCTAssertEqual(255, results[0].hr)
        XCTAssertFalse(results[0].sensorContact)
        XCTAssertFalse(results[0].sensorContactSupported)
        XCTAssertEqual(0, results[0].energy)
        XCTAssertEqual(0, results[0].rrs.count)
        XCTAssertEqual(127, results[1].hr)
        XCTAssertFalse(results[1].sensorContact)
        XCTAssertFalse(results[1].sensorContactSupported)
        XCTAssertEqual(0, results[1].energy)
        XCTAssertEqual(0, results[1].rrs.count)
    }

    // GIVEN BLE HR Service client receives heart rate value in uint16 data format
    // WHEN heart rate stream is consumed
    // THEN the values are emitted
    func testHrFormatUint16() async throws {
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(2, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        bleHrClient.processServiceData(characteristic, data: Data([0x01, 0x80, 0x80]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x01, 0x7F, 0x7F]), err: 0)

        let results = try await task.value

        XCTAssertEqual(32896, results[0].hr)
        XCTAssertFalse(results[0].sensorContact)
        XCTAssertFalse(results[0].sensorContactSupported)
        XCTAssertEqual(0, results[0].energy)
        XCTAssertEqual(0, results[0].rrs.count)
        XCTAssertEqual(32639, results[1].hr)
        XCTAssertFalse(results[1].sensorContact)
        XCTAssertFalse(results[1].sensorContactSupported)
        XCTAssertEqual(0, results[1].energy)
        XCTAssertEqual(0, results[1].rrs.count)
    }

    // GIVEN BLE HR Service client receives data
    // WHEN sensor contact is supported
    // THEN sensor contact corresponds the support flag
    func testSensorContactSupported() async throws {
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(2, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        bleHrClient.processServiceData(characteristic, data: Data([0x06, 0xFF]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x04, 0xFF]), err: 0)

        let results = try await task.value

        XCTAssertTrue(results[0].sensorContact)
        XCTAssertTrue(results[0].sensorContactSupported)
        XCTAssertFalse(results[1].sensorContact)
        XCTAssertTrue(results[1].sensorContactSupported)
    }

    // GIVEN BLE HR Service client receives data
    // WHEN sensor contact is not supported
    // THEN sensor contact shall be always false
    func testSensorContactNotSupported() async throws {
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(2, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        bleHrClient.processServiceData(characteristic, data: Data([0x02, 0x00]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x00, 0x00]), err: 0)

        let results = try await task.value

        XCTAssertFalse(results[0].sensorContact)
        XCTAssertFalse(results[0].sensorContactSupported)
        XCTAssertFalse(results[1].sensorContact)
        XCTAssertFalse(results[1].sensorContactSupported)
    }

    // GIVEN BLE HR Service client receives data
    // WHEN Energy Expended field is present
    // THEN correct energy value is emitted
    func testEnergyExpended() async throws {
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(2, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        bleHrClient.processServiceData(characteristic, data: Data([0x09, 0x00, 0x00, 0xFF, 0xFF]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x08, 0x00, 0x7F, 0x80]), err: 0)

        let results = try await task.value

        XCTAssertEqual(65535, results[0].energy)
        XCTAssertEqual(32895, results[1].energy)
    }

    // GIVEN BLE HR Service client receives data
    // WHEN RR-Interval field is present
    // THEN correct rr values are emitted
    func testRRinterval() async throws {
        let characteristic = CBUUID(string: "2a37")
        let stream = bleHrClient.observeHrNotifications(true)
        let task = Task { try await self.collect(3, from: stream) }

        try await Task.sleep(nanoseconds: 20_000_000)

        bleHrClient.processServiceData(characteristic, data: Data([0x10, 0x00, 0xFF, 0xFF]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x11, 0x00, 0x00, 0xFF, 0xFF, 0x7F, 0x80]), err: 0)
        bleHrClient.processServiceData(characteristic, data: Data([0x10, 0x00, 0x00, 0x00]), err: 0)

        let results = try await task.value

        XCTAssertEqual(3, results.count)

        XCTAssertEqual(0, results[0].hr)
        XCTAssertEqual(1, results[0].rrs.count)
        XCTAssertEqual(1, results[0].rrsMs.count)
        XCTAssertTrue(results[0].rrPresent)
        XCTAssertEqual(65535, results[0].rrs[0])
        XCTAssertEqual(63999, results[0].rrsMs[0])

        XCTAssertEqual(0, results[1].hr)
        XCTAssertEqual(2, results[1].rrs.count)
        XCTAssertEqual(2, results[1].rrsMs.count)
        XCTAssertTrue(results[1].rrPresent)
        XCTAssertEqual(65535, results[1].rrs[0])
        XCTAssertEqual(63999, results[1].rrsMs[0])
        XCTAssertEqual(32895, results[1].rrs[1])
        XCTAssertEqual(32124, results[1].rrsMs[1])

        XCTAssertEqual(0, results[2].hr)
        XCTAssertEqual(1, results[2].rrs.count)
        XCTAssertEqual(1, results[2].rrsMs.count)
        XCTAssertTrue(results[2].rrPresent)
        XCTAssertEqual(0, results[2].rrs[0])
        XCTAssertEqual(0, results[2].rrsMs[0])
    }
}
